### ISSUE TITLE
[Bug Report] ttnn.mean op - Data Mismatch 

### ISSUE THREAD

**Issue by chandrasekaranpradeep** [tenstorrent/tt-metal#13621]
**Describe the bug**
The ttnn.mean throws assertion error because of data mismatch between PyTorch and TTNN output and the pcc is dropped to 0.72 when the input tensor of (1, 12, 3200) and dim = -1 is passed to  ttnn.mean op.
For more context, here is the exact error message
```
def assert_with_pcc(expected_pytorch_result, actual_pytorch_result, pcc=0.9999):
        assert list(expected_pytorch_result.shape) == list(
            actual_pytorch_result.shape
        ), f"list(expected_pytorch_result.shape)={list(expected_pytorch_result.shape)} vs list(actual_pytorch_result.shape)={list(actual_pytorch_result.shape)}"
        pcc_passed, pcc_message = comp_pcc(expected_pytorch_result, actual_pytorch_result, pcc)
>       assert pcc_passed, construct_pcc_assert_message(pcc_message, expected_pytorch_result, actual_pytorch_result)
E       AssertionError: 0.7203957195745748
```

**To Reproduce**

Run the following test:
```
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc
from models.utility_functions import torch_random
def test_mean_pcc_issue(device):
    torch.manual_seed(0)

    input_shape = (1, 12, 3200)
    reduce_dim = -1
    
    torch_input_tensor = torch.rand(input_shape, dtype=torch.float32)
    torch_output_tensor = torch.mean(torch_input_tensor, dim=reduce_dim, keepdim=True, dtype=torch.float32)

    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
    
    output_tensor = ttnn.mean(input_tensor, dim=reduce_dim)
    output_tensor = ttnn.to_torch(output_tensor)
    
    assert_with_pcc(torch_output_tensor, output_tensor)
```

**Expected behavior**
The data mismatch between PyTorch and TTNN output should be resolved.

**Comment by sdjordjevicTT** [comment]
@ntarafdar @sjameelTT can you please help me to find owners for this issue?

**Comment by ntarafdar** [comment]
hey @sdjordjevicTT asking around, its a reduction op who doesn't have an owner , will ask ttnn ppl and get back to you

**Comment by ntarafdar** [comment]
@sdjordjevicTT asked around and since there is no other owner for this, the TMs team will have to take this. 
We cannot get to this until end of next week.

**Comment by sdjordjevicTT** [comment]
Thanks @ntarafdar for picking this up. Great, I believe that should work for us.

**Comment by sdjordjevicTT** [comment]
@nvukobratTT can comment more about priority, but I think this issue blocks Llama 3B bring-up on the Forge side.

**Comment by nvukobratTT** [comment]
> moving to a P1 issue. @sdjordjevicTT pls comment if you believe the P0 is justified.

Confirming what @sdjordjevicTT mentioned, this one is a blocker for the Open Llama 3B model. 

Additional details can be found on the MLIR issue as well:
- https://github.com/tenstorrent/tt-mlir/issues/869

**Comment by ntarafdar** [comment]
Spoke to Jasmine, and @bbradelTT is for now taking over reductions. I'm reassigning this to him.

**Comment by bbradelTT** [comment]
I tried to find out if there's any point at which there's a big drop off. Seemed like it might be somewhere between 1200 and 1400, but the PCC value goes up and down a fair amount:
```
    #input_shape = (1, 12, 3200) # .72
    #input_shape = (1, 12, 1600) # .76
    #input_shape = (1, 12, 1400) # .78
    #input_shape = (1, 12, 1376) # .81
    #input_shape = (1, 12, 1363) # .93
    #input_shape = (1, 12, 1369) # .83
    #input_shape = (1, 12, 1368) # .85
    #input_shape = (1, 12, 1367) # .71
    #input_shape = (1, 12, 1366) # .96
    #input_shape = (1, 12, 1350) # .95
    #input_shape = (1, 12, 1344) # .87
    #input_shape = (1, 12, 1300) # .91
    #input_shape = (1, 12, 1200) # .93
    #input_shape = (1, 12, 800) # .92
    #input_shape = (1, 12, 320) # .99
```

**Comment by sdjordjevicTT** [comment]
Hi @bbradelTT do we have some updates regarding this missmatch problem?

**Comment by bbradelTT** [comment]
@sdjordjevicTT Unfortunately we need to overhaul reduce. I won't have concrete updates for a while.

**Comment by nvukobratTT** [comment]
> [@sdjordjevicTT](https://github.com/sdjordjevicTT) Unfortunately we need to overhaul reduce. I won't have concrete updates for a while.

@bbradelTT thanks for the details. Can you clarify the following:
1. What are the core issues with reduced mean and related PCC issues? 
    - Having this info, we might be able to work around it until a fix is in place
1. Details around lowering the priority on this issue.
    - From the current standpoint, this issue should be treated as P0 as it blocks one of the Forge core models, Llama.
  
To be certain that this issue is properly tracked, I'm re-adding the P0 label once again. Please correct me if I'm missing some context as to why this one should still be a P1 issue. 

Thanks!

**Comment by bbradelTT** [comment]
@nvukobratTT 
1. At a high level, depending on the inputs, a lot of different things are done (including transpose, auto format, and reshape) that don't work that well when different dimensions are padded, and padding may not be done properly in all instances (see https://github.com/tenstorrent/tt-metal/issues/12662 and https://github.com/tenstorrent/tt-metal/issues/13647). I'm not sure of the root cause of the reduced mean in this case, although in one of the other issues it seems to be because the entire tile is used and therefore there are extra 0s and the denominator is larger than it should be. Having said that, I still haven't isolated the exact issue for this specific scenario.
2. I don't have context for the priority.

Update for today:
- I'm working on printing out the data in the circular buffers to see where any discrepancies are occurring. Will continue that work tomorrow.

**Comment by nvukobratTT** [comment]
Thanks for pushing this one further @bbradelTT! Much appreciated 🙌

**Comment by bbradelTT** [comment]
Update for today:
- I spent today in meetings and on my other currently active P0 that I didn't have a chance to look at yesterday.

**Comment by nvukobratTT** [comment]
> Update for today:
> 
> * I spent today in meetings and on my other currently active P0 that I didn't have a chance to look at yesterday.

Thanks for the update and for letting us know Borys :)) 

It's valuable for us to know the state of the issues, and when we expect it to be resolved, so that we can plan accordingly on our side as well. Thanks once again!

**Comment by bbradelTT** [comment]
Update for today:
- instrumented compute kernel to print out input CBs and output CB and used all ones as input.
- Used ones instead of random values.
- There is a scalar factor that is 1/n. When n is not a power of 2, there is error introduced.
- The reduce_tile operation is called for each tile of the reduced operation, multiplying the error.

Looking at different shapes:
- 1,1,12,32 scalar is exact, output is 1
- 1,1,12,800 scalar is 0.00124359 vs .00125 (.6% off), output is 0.988281
- 1,1,12,3200 scalar is 0.000310898 vs 0.0003125 (.5% off), output is 0.917969
- 1,1,12,2048 scalar is exact, output is 1

Next steps:
As far as I can tell, the floating point errors from reduce_tile are accumulative. Therefore the next step is to try to remove the scaling factor (leave it as 1) from reduce_tile and apply it afterwards at the end. I will need to analyze how that helps PCC. That will have a performance hit, but hopefully will improve PCC.

cc @rtawfik01 for LLK visibility

**Comment by bbradelTT** [comment]
Removing the scaling factor and then calling ttnn::multiply resulted in the following.

With all ones, increased the accuracy:
- 1,1,12,800 output is 0.9961
- 1,1,12,3200 output is 0.9961
- 1,1,12,2048 output is still 1

However, with random inputs, lowered PCC:

```
expected_pytorch_result = tensor([[[0.4979],
         [0.4969],
         [0.5080],
         [0.5029],
         [0.5012],
         [0.5046],
         [0.4993],
         [0.5034],
         [0.5109],
         [0.4984],
         [0.4972],
         [0.4963]]])
actual_pytorch_result = TorchTensor([[[0.4961],
              [0.4980],
              [0.5000],
              [0.4961],
              [0.5039]...            [0.5039],
              [0.5078],
              [0.5000],
              [0.5078],
              [0.4961]]])
pcc = 0.9999

    def assert_with_pcc(expected_pytorch_result, actual_pytorch_result, pcc=0.9999):
        assert list(expected_pytorch_result.shape) == list(
            actual_pytorch_result.shape
        ), f"list(expected_pytorch_result.shape)={list(expected_pytorch_result.shape)} vs list(actual_pytorch_result.shape)={list(actual_pytorch_result.shape)}"
        pcc_passed, pcc_message = comp_pcc(expected_pytorch_result, actual_pytorch_result, pcc)
>       assert pcc_passed, construct_pcc_assert_message(pcc_message, expected_pytorch_result, actual_pytorch_result)
E       AssertionError: 0.3941431853119553
```

Will look at CBs to see what happened.

**Comment by bbradelTT** [comment]
Looking at CBs, they look like the final output is integers. That lead me to think there is an incremental factor at play as well.
```
FAILED test_mean2.py::test_reduce[32] - AssertionError: 0.9996477335026296
FAILED test_mean2.py::test_reduce[50] - AssertionError: 0.9990662794549972
FAILED test_mean2.py::test_reduce[128] - AssertionError: 0.9983768209716934
FAILED test_mean2.py::test_reduce[200] - AssertionError: 0.9971359814011025
FAILED test_mean2.py::test_reduce[512] - AssertionError: 0.9640920022989882
FAILED test_mean2.py::test_reduce[800] - AssertionError: 0.9159602157408702
FAILED test_mean2.py::test_reduce[1600] - AssertionError: 0.8340448198726398
FAILED test_mean2.py::test_reduce[3200] - AssertionError: 0.3941431853119553
FAILED test_mean2.py::test_reduce[4096] - AssertionError: 0.752128086868872
```

There is a definite drop off, although powers of 2 have different behaviour.

Looking at the code, the mean reduce on the width (last dimension) is a matmul_tiles call, not a reduce_tile call.

I swapped dimensions, and verified that this uses reduce_tile, and that PCC behaviour is similar:
```
def test_reduce_h(device, width):
    torch.manual_seed(0) 
    
    input_shape = (width, 12)
    reduce_dim = 0
    
    torch_input_tensor = torch.rand(input_shape, dtype=torch.float32)
    torch_output_tensor = torch.mean(torch_input_tensor, dim=reduce_dim, keepdim=True, dtype=torch.float32)
    
    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)

    output_tensor = ttnn.mean(input_tensor, dim=reduce_dim)
    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_output_tensor, output_tensor)
```
resulted in
```
FAILED test_mean2.py::test_reduce_h[32] - AssertionError: 0.9998316111876924
FAILED test_mean2.py::test_reduce_h[50] - AssertionError: 0.999309887656723
FAILED test_mean2.py::test_reduce_h[128] - AssertionError: 0.9949648376563645
FAILED test_mean2.py::test_reduce_h[200] - AssertionError: 0.9940303932238288
FAILED test_mean2.py::test_reduce_h[512] - AssertionError: 0.9832944977683854
FAILED test_mean2.py::test_reduce_h[800] - AssertionError: 0.9470958968139239
FAILED test_mean2.py::test_reduce_h[1600] - AssertionError: 0.7357217229341707
FAILED test_mean2.py::test_reduce_h[3200] - AssertionError: 0.7696898216846108
FAILED test_mean2.py::test_reduce_h[4096] - AssertionError: 0.6391112138824813
```

I think what might be happening is that the logic is processing one tile at a time, accumulating the numbers.

On average each value would be around .5. By the time you get to a width of 1600, the sum would be 800. Adding .5 to 800 would result in more precision loss.
That would also explain why the PCC went down faster when the multiplication is done at the end. The accumulated values are larger.

Next step: looking at reducing in more of an `n*log(n)` manner in the pytest to see if that helps PCC.

**Comment by bbradelTT** [comment]
I had to switch to reduce sum, which is independent of additional 0s being padded in because of fatals in padding.

PCCs follow a similar trend:
```
FAILED test_mean2.py::test_reduce[32] - AssertionError: 0.9996477335026296
FAILED test_mean2.py::test_reduce[50] - AssertionError: 0.9992098215952638
FAILED test_mean2.py::test_reduce[128] - AssertionError: 0.9983768209716934
FAILED test_mean2.py::test_reduce[200] - AssertionError: 0.9980993887635425
FAILED test_mean2.py::test_reduce[512] - AssertionError: 0.9640920022989882
FAILED test_mean2.py::test_reduce[800] - AssertionError: 0.9242553618875884
FAILED test_mean2.py::test_reduce[1600] - AssertionError: 0.8777743499163675
FAILED test_mean2.py::test_reduce[3200] - AssertionError: 0.3992921039572661
FAILED test_mean2.py::test_reduce[4096] - AssertionError: 0.752128086868872
FAILED test_mean2.py::test_reduce[32768] - AssertionError: 0.03179110579730596
```

Adding in chunks does help PCC
```
def test_reduce_part(device):
    torch.manual_seed(0)
        
    # 32, 1024, 32768
    width = 1024
    input_shape = (12, width)
    reduce_dim = -1

    n = 32
    torch_input_tensors = [torch.rand(input_shape, dtype=torch.float32) for i in range(n)]
    torch_input_tensor = torch.cat(torch_input_tensors, reduce_dim)
    
    torch_output_tensor = torch.sum(torch_input_tensor, dim=reduce_dim, keepdim=True, dtype=torch.float32)
    
    input_tensors = [ttnn.from_torch(torch_input_tensors[i], dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device) for i in range(n)]
    
    intermediate_tensors = [ttnn.sum(input_tensors[i], dim=reduce_dim) for i in range(n)]
    intermediate_torch_tensors  = [ttnn.to_torch(intermediate_tensors[i]) for i in range(n)]
    intermediate_torch_tensor = torch.concat(intermediate_torch_tensors, reduce_dim)
    intermediate_tensor = ttnn.from_torch(intermediate_torch_tensor, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
    output_tensor = ttnn.sum(intermediate_tensor, dim=reduce_dim)
    output_tensor = ttnn.to_torch(output_tensor)
    
    assert_with_pcc(torch_output_tensor, output_tensor)
```
does improve PCC:
```
32768 (width = 1024)
FAILED test_mean2.py::test_reduce_part - AssertionError: 0.7728428816004724
8192 (width = 256)
FAILED test_mean2.py::test_reduce_part - AssertionError: 0.9844695593902897
```

I also looked online.

Loss of precision is a known problem for adding long sequences of floating point numbers, and there are algorithms to help with that:
https://en.wikipedia.org/wiki/Kahan_summation_algorithm

Next step: I'll see how complex it would be to try to incorporate such an algorithm here.

**Comment by bbradelTT** [comment]
Another algorithm is https://en.wikipedia.org/wiki/Pairwise_summation which is what I tried to approximate with the adjusted test.

**Comment by nvukobratTT** [comment]
@bbradelTT, thanks for pushing this further and keeping us posted! 🙌

**Comment by bbradelTT** [comment]
Today's update:
- I was working on other high priority work
- should have some time tomorrow to work on this

**Comment by bbradelTT** [comment]
Today's update:
- I thought #13647 was related and looked into how the two are related. Turns out there were unrelated issues, and I created a PR. Will continue to work on this tomorrow.

**Comment by bbradelTT** [comment]
Today's update:
- started implementation
- should be able to get an initial PR out early next week

**Comment by bbradelTT** [comment]
Today's update:
- initial implementation is done. PCC goes up to .978 for size of 3200
- there are still some errors that I need to debug, and I don't know how much higher the PCC can go

**Comment by bbradelTT** [comment]
Still one issue with pad where inputs are not multiples of tiles. However, I got the following PCCs for inputs of size (1, 1, 12, n) where n is passed in:
```
FAILED test_mean2.py::test_reduce[32] - AssertionError: 0.9996477335026296
FAILED test_mean2.py::test_reduce[50] - AssertionError: 0.9992098215952638
FAILED test_mean2.py::test_reduce[128] - AssertionError: 0.9990462253048942
FAILED test_mean2.py::test_reduce[512] - AssertionError: 0.9965267965250721
FAILED test_mean2.py::test_reduce[1024] - AssertionError: 0.9889842589831356
FAILED test_mean2.py::test_reduce[1600] - AssertionError: 0.9852341708352339
FAILED test_mean2.py::test_reduce[3200] - AssertionError: 0.9786824431459795
FAILED test_mean2.py::test_reduce[4096] - AssertionError: 0.963620489697217
FAILED test_mean2.py::test_reduce[32768] - AssertionError: 0.7968057244607337
```

Would this be sufficient?

**Comment by nvukobratTT** [comment]
Thanks for the update @bbradelTT! 

@dgolubovicTT can you confirm if we need confirmation for other shapes besides the `(1, 1, 12, n)` for the reduce mean? 

@bbradelTT for these PCC issues to be generally fixed on the TTNN side, do you have a separate issue? Even if the fix for this case is currently OK, I assume we will hit many more of them pretty soon..

Thanks once again for helping us out with this one!

**Comment by dgolubovicTT** [comment]
@nvukobratTT 12 probably comes from initial nuber of tokens in llama. @chandrasekaranpradeep correct me if I am wrong, but since number of input tokens can be any number less than max input tokens => we need ttnn.mean to work generally (not just 12).

**Comment by chandrasekaranpradeep** [comment]
@dgolubovicTT Thats Right
The reduce op input shape is (1, 12, 3200)
Here, 
   1 - refers to batch_size
   12 -> refers to number of token in the sequences (It will vary any number less than max sequence length)
   3200 -> embedding dimension (Fixed for open llama)
As @dgolubovicTT  mentioned,  we need ttnn.mean to work generally

**Comment by dgolubovicTT** [comment]
@chandrasekaranpradeep I would suggest you add test for this on our side that will cover all input shapes from (1,1,3200) to (1,max_num_tokens, 3200). Therefore, we will know for sure what shapes fail. This is overlapping with unique ops testing, and I think this is a good example for how to use it. 
We have to support all ops in llama that have variable sequence length for all sequence lengths, right?
So we should add those tests covering all sequence lengths as soon as possible, for two reasons:
1. Prevent regressions (catch them early)
2. See how far we are from generality on each ttnn op with respect to sequence length.
What are your thoughts @chandrasekaranpradeep @nvukobratTT. We can resume this topic offline, to avoid overloading this thread...

**Comment by dgolubovicTT** [comment]
Since inputs are from torch.rand (which is Uniform distribution [0,1)) it is really odd that mean over 3200 values is 0.46-0.48. It seems that mean doesn't work on bigger dimensions at all...

**Comment by bbradelTT** [comment]
Today's update:
- still blocked by pad. Created https://github.com/tenstorrent/tt-metal/issues/15511 to track fixing pad

**Comment by dgolubovicTT** [comment]
Hey @bbradelTT,
I've talked to @nvukobratTT offline to scope requirements for this issue.

For the first phase we need ttnn.mean to work for shapes (1,12,3200), (1,128,3200) and (1,1,3200). Fixing this is P0.

For the second phase, we will need ttnn.mean to work for all shapes (1,x,3200) for x in [1,2048]. Only then, we can get Llama inference for any number of tokens on input. This second phase can be P1. 

So after you finish this first phase, we will move this issue to P1, and go from there.

**Comment by bbradelTT** [comment]
Update:
- still blocked

I created https://github.com/tenstorrent/tt-metal/issues/15656 to look at improving accuracy further.

**Comment by bbradelTT** [comment]
Update:
- accumulating into FP32 brought the PCC high enough that the tree add algorithm is accurate enough.

**Comment by bbradelTT** [comment]
@dgolubovicTT @nvukobratTT
 
As a temporary workaround, could you try using a compute kernel with fp32 acc enabled and let me know if there is a change in PCC?

Something like
```
    compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=False, # or True, doesn't matter
        )
output_tensor = ttnn.mean(input_tensor, dim=reduce_dim, compute_kernel_config=compute_kernel_config)
```

Update:
- still blocked
- to help with PCC tried to enable fp32_dest_acc_en by default if compute_kernel_config is not specified but can't because of an assert due to an issue on BH that hasn't been fixed yet

**Comment by dgolubovicTT** [comment]
@bbradelTT is this workaround for shape (1,12,3200) or it should make all above shapes work? Also, to test some workaround I would need tt-metal commit that I can cherry-pick and not just a code snippet...

Can you tell me which of above mentioned critical shapes now works on tt-metal main branch, if any ((1,12,3200), (1,128,3200) and (1,1,3200))?

**Comment by dgolubovicTT** [comment]
To explain, we don't need ttnn.mean to just work in ttnn test, so workaround that uses compute_kernel_config in ttnn doesn't fix the problem...

**Comment by bbradelTT** [comment]
> [@bbradelTT](https://github.com/bbradelTT) is this workaround for shape (1,12,3200) or it should make all above shapes work?

It should improve PCC for all shapes. However, I don't know by how much, and right now I would just like to know about (1,12,3200). 

Unfortunately I have a lot of edits on my branch, which means I can't test this change in isolation. I'm hoping that this is just a python change you would be able to test it easily on your end.

**Comment by dgolubovicTT** [comment]
@bbradelTT I tried this out and it fixes PCC fully. We have an option to call ttnn.mean with this `compute_kernel_config` in runtime. Does this `compute_kernel_config` have any perf repercussions? We would use it for each call of ttnn.mean so it is important to know this.

**Comment by bbradelTT** [comment]
It, has perf repercussions, but they should be small, and I'm going to need to add that in if you don't specify a compute_kernel_config. 

Which means unless you specify a compute kernel config with that off, you'll see the perf impact.

Also, it'll have a lot less perf repercussions than what I'm working on, which is probably necessary since GS doesn't have fp32.

**Comment by ayerofieiev-tt** [comment]
Subscribing

**Comment by bbradelTT** [comment]
Update:
- still blocked on underlying issues
- I looked at behaviour on GS of my tree add change against baseline on GS (where fp32 is not on option and got the following)

```
@pytest.mark.parametrize("input_shape", [
    (1, 12, 3200),
    (1, 1, 3200),
    (1, 128, 3200),
    (1, 2048, 3200),
    ])
def test_large_tensor_mean(device, input_shape):
    reduce_dim = -1
    torch_input_tensor = torch.rand(input_shape, dtype=torch.float32)
    torch_output_tensor = torch.mean(torch_input_tensor, dim=reduce_dim, keepdim=True, dtype=torch.float32)

    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    output_tensor = ttnn.mean(input_tensor, dim=reduce_dim)
    output_tensor = ttnn.to_torch(output_tensor)
    assert_with_pcc(torch_output_tensor, output_tensor)
```
produces
```
on GS: 
Base:
PASSED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 1, 3200)]
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 12, 3200)] - AssertionError: 0.7769184620303734
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 128, 3200)] - AssertionError: 0.7137656700429718
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 2048, 3200)] - AssertionError: 0.6815292219338761

with cut at 64:
PASSED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 1, 3200)]
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 12, 3200)] - AssertionError: 0.9772472040616764
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 128, 3200)] - AssertionError: 0.9666487764829472
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 2048, 3200)] - AssertionError: 0.9527421326781462

with cut at 128:
PASSED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 1, 3200)]
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 12, 3200)] - AssertionError: 0.9335399543286304
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 128, 3200)] - AssertionError: 0.9003435192493461
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 2048, 3200)] - AssertionError: 0.8903827326146612

with cut at 256:
PASSED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 1, 3200)]
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 12, 3200)] - AssertionError: 0.9522780737237543
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 128, 3200)] - AssertionError: 0.9058280173941267
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 2048, 3200)] - AssertionError: 0.8969139544477592

with cut at 512:
PASSED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 1, 3200)]
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 12, 3200)] - AssertionError: 0.9544566782083125
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 128, 3200)] - AssertionError: 0.8692186521807318
FAILED tests/ttnn/unit_tests/operations/test_reduction_mean.py::test_large_tensor_mean[input_shape=(1, 2048, 3200)] - AssertionError: 0.8555394079497131
```

I'm leaning towards cutting tensors at every 64 elements (2 tiles), but that will have a larger perf impact.

**Comment by bbradelTT** [comment]
For larger tensors, the tree add algorithm changes accuracy on GS from 0.7 to 0.85-0.95. However, there is going to be a pretty big performance hit for mean/sum/std/var.

The options are:
1. Just enable fp32 dest acc by default, which will only help WH onwards
2. Enable fp32 dest acc by default and add in the tree add algorithm while dividing the tensors at a fine granularity
3. Enable fp32 dest acc by default and add in the tree add algorithm while dividing the tensors at a course granularity

@davorchap 
How much do we care about accuracy on GS?
Do you have a preference for which option I go with?

**Comment by bbradelTT** [comment]
Update:
- still waiting for underlying issues to be fixed
- enabled fp32 dest acc, will hopefully help with PCC on WH at least

**Comment by ntarafdar** [comment]
update on one of the underlying issues, [pad last dim](https://github.com/tenstorrent/tt-metal/issues/15511)
@jaykru-tt is doing a good job in rewriting the kernel from scratch to handle misaligned read/writes. We expect a solution early to mid next week .

**Comment by dgolubovicTT** [comment]
I've rebased onto [tt-metal main ](https://github.com/tenstorrent/tt-metal/commit/60373c88f423718b2a87031ac87e502fd2af4d0e) and rerun tests for shape (1,x,3200) where x goes from 1 to 2048. PCCs are sufficiently good without the use of the workaround (compute_kernel_config). Thanks for prioritizing this guys! Closing this one
